networkx/algorithms/cluster.py

```diff
@@ -47,17 +49,42 @@ def triangles(G, nodes=None):
 
     Notes
     -----
-    When computing triangles for the entire graph each triangle is counted
-    three times, once at each node.  Self loops are ignored.
+    Self loops are ignored.
 
     """
     # If `nodes` represents a single node in the graph, return only its number
     # of triangles.
     if nodes in G:
         return next(_triangles_and_degree_iter(G, nodes))[2] // 2
-    # Otherwise, `nodes` represents an iterable of nodes, so return a
+
+    # if `nodes` is a container of nodes, then return a
     # dictionary mapping node to number of triangles.
-    return {v: t // 2 for v, d, t, _ in _triangles_and_degree_iter(G, nodes)}
+    if nodes:
```

Comment:
Ah, interesting, didn't consider this! 

If `nodes=0` and G contains this node, then we would get back result from line 58. If `nodes=0` and G does not contain this node, without implementing `if nodes is not None`, the rest of the code would execute, triggering unnecessary compute. 

The tricky part is that if we change line 62 to `if nodes is not None`, then the returned value is going to be a `dict`, while the user would probably expect an `int` (since `nodes` is a single node `0`).

Another tricky example is if nodes is set to `False`, which could also be a valid node, but for which `if nodes` is not a valid solution.

So, I'm not sure if this is a good way out of this, but how about checking whether adding a check whether `nodes` could refer to a single node (by checking if it is hashable), if it is, then return the triangle count (or 0 if the node is not in the graph). Kindly check the updated version.

Reply:
This idiom appears often in our code, and we sometimes refer to it as a `nbunch`. It could be a node. If could be a collection of nodes, and it could be None which means all the nodes. Because nodes can be containers, like the grid coordinates: (2, 3), we need to handle them first. If (2, 3) is in G, then we will treat it as a single node. And if (2, 3) is not in G then we will treat it as a container. This same logic works for non-containers too. We first check if it is in G. If not, then we check if it is None and handle the resulting container appropriately. The old `nx.triangle` code does this too. So we can leave some of that from before if that's easier.

```python
if nodes in G:
    return singleton  # nodes is a node of the graph G

if nodes is not None:
    #  nodes must be an iterator of nodes, so handle that case
    return dict(...)

# build the dict for all nodes in G
```

Reply:
Maybe I'm thinking of very unrealistic edge cases, but consider the snippet below:

```python
from networkx import path_graph, triangles

G = path_graph(5)
target_node = (3,4)

# compute triangles for the target node if it's not part of a graph (expect 0)
print(triangles(G, nodes=target_node))
# {3: 0, 4: 0}

# compute triangles for the target node if it's part of a graph (expect 0)
G.add_node(target_node)
print(triangles(G, nodes=(3, 4)))
# 0
```

For me the behaviour is potentially ambiguous. Specifically, in the first case, if there's an explicit check on the hashability of the object passed to `nodes`, then we would have received 0, since the code would check whether `target_node` is part of the graph.

I think this is a larger question about flexibility/ambiguity of `nbunch`, perhaps there should be a separate class for `nbunch`... For now, I think having an explicit check on hashability is going to resolve this edge case.

Reply:
Checking hashability doesn't solve the issue because strings and tuples are iteraables that are hashable (and frozensets, frozen networkx.Graphs, etc).

Reply:
Hmm, from the [tutorial](https://networkx.org/documentation/stable/tutorial.html): "a node can be any hashable object (except `None`)". Hence, if we check that an object is hashable, it could potentially be a node and so could be tested with `some_object in G`. So, strings/tuples/etc can be a node and when passed as an individual object to `nodes` they will be interpreted as a node, rather than as a container of nodes. I will elaborate in the issue on `nbunch`.